Implementation for HRINT-1503

myarichuk opened the pull request:
*  Introduce single-threaded execution of tests with deterministic alphabetic ordering for DEBUG, VALIDATE or TEST_PERFORMANCE compilation variables
* wire test method beginning and ending via xUnit to add snapshots of memory/cpu via TestResourcesAnalyzer
* add perf test assembly runner to FastTests, SlowTests and RachisTests

myarichuk committed:
Introduce single-threaded execution of tests with deterministic alphabetic ordering for DEBUG, VALIDATE or TEST_PERFORMANCE compilation variables - WIP (HRINT-1503)

myarichuk committed:
Two changes:
* wire test method beginning and ending via xUnit to add snapshots of memory/cpu via TestResourcesAnalyzer
* add perf test assembly runner to SlowTests and RachisTests

ppekrol commented on .gitignore:
```diff
@@ -122,3 +122,5 @@ bench/Micro.Benchmark/BenchmarkDotNet.Artifacts/
 /src/Raven.Pal/*.nativecodeanalysis.xml
 /src/Raven.Pal/src/posix/*.o
+/src/Raven.Pal/Debug
+/src/Raven.Pal/build-all-posix.log
```
Is this needed?

myarichuk replied:
No. I will delete those, it should be in .gitignore

ppekrol commented on test/SlowTests/SlowTests.csproj:
```diff
@@ -32,6 +32,7 @@
     <None Remove="Data\RavenDB_13940\storage-with-reused-journal-and-synced-data.zip" />
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="..\Tests.Infrastructure\CommonTestsAssemblyInfo.cs" Link="Properties\CommonTestsAssemblyInfo.cs" />
```
Don't forget about Embedded

ppekrol commented on test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using Xunit;
+
+#if DEBUG || VALIDATE || TEST_PERFORMANCE
```
only `TESTING`

myarichuk replied:
Ok. I will remove the DEBUG from there once everything else will be in place. No?

ppekrol commented on test/Tests.Infrastructure/CommonTestsAssemblyInfo.cs:
```diff
@@ -0,0 +1,10 @@
+
+[assembly: TestCaseOrderer("Tests.Infrastructure.AlphabeticTestsOrderer", "Tests.Infrastructure")]
+[assembly: TestCollectionOrderer("Tests.Infrastructure.AlphabeticTestsOrderer", "Tests.Infrastructure")]
+[assembly: CollectionBehavior(DisableTestParallelization = true)]
```
Let's control that via xunit.setting.json

myarichuk replied:
I have removed the DisableTestParallelization, as we discussed.

ppekrol commented on test/Tests.Infrastructure/AlphabeticTestsOrderer.cs:
```diff
@@ -0,0 +1,34 @@
+using Xunit.Sdk;
+using DiagnosticMessage = Xunit.Sdk.DiagnosticMessage;
+
+namespace Tests.Infrastructure
```
Move everything under one namespace

ppekrol commented on test/Tests.Infrastructure/TestResourcesAnalyzer.cs:
```diff
@@ -33,6 +35,21 @@ static TestResourcesAnalyzer()
+                _resources.Enqueue(TestResources.Create(GetDisplayName(testMethod), nameof(End)));
+        }
+
         internal static void Start(Context context)
```
Need a cleanup here. We do not want to keep all the methods. Also we want to make it synchronous.

myarichuk replied:
I will refactor this to be sync

myarichuk replied:
Done (in the next commit)

ppekrol commented on test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs:
```diff
@@ -0,0 +1,38 @@
+            return base.RunTestMethodAsync(testMethod, method, testCases, constructorArguments)
+                       .ContinueWith(t =>
+                       {
+                           TestResourcesAnalyzer.End(testMethod);
```
Remember that we want to gather 2 results here. Before and after GC.

myarichuk replied:
Done (in the next commit)

myarichuk committed:
Changes after code review + refactor TestResourceAnalyzer to be sync + refine events it writes (HRINT-1503)

myarichuk committed:
refactoring for TestResourceSnapshotWriter (more fields), add Tests.ResourceSnapshotAggregator - a service to receive notifications from Jenkins and write the CSV from artifacts in both RavenDB and InfluxDB (this is still WIP, the commit is for code review)

ppekrol commented on test/Tests.Infrastructure/XunitExtensions/PerfTestClassRunner.cs:
```diff
@@ -0,0 +1,76 @@
+                           if(_resourceSnapshotEnabled)
+                               _testResourceSnapshotWriter.WriteResourceSnapshot(TestStage.TestFinishedBeforeGc, GetDisplayName(testMethod));
+
+                           GC.Collect(2, GCCollectionMode.Forced, true, true);
```
Do we want to compact LOH?
Also do we need to GC.WaitForPendingFinalizers?

myarichuk replied:
I think yes. Added that in the next commit.

ppekrol reviewed (commented):
LGTM

myarichuk commented:
Added an implementation of the service, it should do inserts to InfluxDB. 
* Note the **readme.txt** - for convenience, it contains docker commands to set up local environment (influx + influx UI + grafana)
* Note the *InfluxTag* attribute on the row object - it allows InfluxDB to group-by on that field

myarichuk committed:
* Multiple changes for HRINT-1503:
-> Proper import of resource snapshot data into InfluxDB
-> Calculate deltas between test/end beginning when inserting into InfluxDB
-> Added endpoint to upload CSVs of test resource snapshots
-> Refactored test snapshot resource writers
-> Readme file with Docker commands to set up local environment
